channels/geometry/client/geometry_main.c

```diff
@@ -0,0 +1,365 @@
+/**
+ * FreeRDP: A Remote Desktop Protocol Implementation
+ * Geometry tracking Virtual Channel Extension
+ *
+ * Copyright 2017 David Fort <contact@hardening-consulting.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <winpr/crt.h>
+#include <winpr/synch.h>
+#include <winpr/print.h>
+#include <winpr/stream.h>
+#include <winpr/cmdline.h>
+#include <winpr/collections.h>
+
+#include <freerdp/addin.h>
+#include <freerdp/client/geometry.h>
+#include <freerdp/channels/log.h>
+
+#define TAG CHANNELS_TAG("geometry.client")
+
+#include "geometry_main.h"
+
+struct _GEOMETRY_CHANNEL_CALLBACK
+{
+	IWTSVirtualChannelCallback iface;
+
+	IWTSPlugin* plugin;
+	IWTSVirtualChannelManager* channel_mgr;
+	IWTSVirtualChannel* channel;
+};
+typedef struct _GEOMETRY_CHANNEL_CALLBACK GEOMETRY_CHANNEL_CALLBACK;
+
+struct _GEOMETRY_LISTENER_CALLBACK
+{
+	IWTSListenerCallback iface;
+
+	IWTSPlugin* plugin;
+	IWTSVirtualChannelManager* channel_mgr;
+	GEOMETRY_CHANNEL_CALLBACK* channel_callback;
+};
+typedef struct _GEOMETRY_LISTENER_CALLBACK GEOMETRY_LISTENER_CALLBACK;
+
+struct _GEOMETRY_PLUGIN
+{
+	IWTSPlugin iface;
+
+	IWTSListener* listener;
+	GEOMETRY_LISTENER_CALLBACK* listener_callback;
+
+	GeometryClientContext *context;
+};
+typedef struct _GEOMETRY_PLUGIN GEOMETRY_PLUGIN;
+
+
+static UINT32 geometry_read_RGNDATA(wStream *s, UINT32 len, FREERDP_RGNDATA *rgndata)
+{
+	UINT32 dwSize, iType;
+	INT32 right, bottom;
+
+	if (len < 32)
+	{
+		WLog_ERR(TAG, "invalid RGNDATA");
+		return ERROR_INVALID_DATA;
+	}
+
+	Stream_Read_UINT32(s, dwSize);
+	if (dwSize != 32) {
+		WLog_ERR(TAG, "invalid RGNDATA dwSize");
+		return ERROR_INVALID_DATA;
+	}
+
+	Stream_Read_UINT32(s, iType);
+	if (iType != RDH_RECTANGLE) {
+		WLog_ERR(TAG, "iType %"PRIu32" for RGNDATA is not supported", iType);
+		return ERROR_UNSUPPORTED_TYPE;
+	}
+
+	Stream_Read_UINT32(s, rgndata->nRectCount);
+	Stream_Seek_UINT32(s); /* nRgnSize IGNORED */
+	Stream_Read_INT32(s, rgndata->boundingRect.x);
+	Stream_Read_INT32(s, rgndata->boundingRect.y);
+	Stream_Read_INT32(s, right);
+	Stream_Read_INT32(s, bottom);
+
+	rgndata->boundingRect.width = right - rgndata->boundingRect.x;
+	rgndata->boundingRect.height = bottom - rgndata->boundingRect.y;
+
+	len -= 32;
+	if (len / 4 < rgndata->nRectCount)
```

Comment:
I think this should be ```len / 16``` (4 * sizeof(UINT32))

Reply:
good catch